From 3754a384c0f13b858bb1a191f4b212c1e3fc6e64 Mon Sep 17 00:00:00 2001
From: Marco Martin <notmart@gmail.com>
Date: Fri, 17 Apr 2020 10:01:59 +0200
Subject: sheet must be under other popups

BUG:419930
---
 src/controls/templates/OverlaySheet.qml | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/controls/templates/OverlaySheet.qml b/src/controls/templates/OverlaySheet.qml
index a9673b0..1b962fe 100644
--- a/src/controls/templates/OverlaySheet.qml
+++ b/src/controls/templates/OverlaySheet.qml
@@ -208,7 +208,6 @@ QtObject {
 
         anchors.fill: parent
 
-        z: 9998
         visible: false
         drag.filterChildren: true
         hoverEnabled: true
-- 
cgit v1.1

