From e7204d54ff3fd82bf7852d7af5040151faede78c Mon Sep 17 00:00:00 2001
From: Marco Martin <notmart@gmail.com>
Date: Fri, 17 Apr 2020 10:16:48 +0200
Subject: never close when the click is inside the sheet contents

BUG:419691
---
 src/controls/templates/OverlaySheet.qml | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/controls/templates/OverlaySheet.qml b/src/controls/templates/OverlaySheet.qml
index 1b962fe..1f4d9e3 100644
--- a/src/controls/templates/OverlaySheet.qml
+++ b/src/controls/templates/OverlaySheet.qml
@@ -214,8 +214,8 @@ QtObject {
         clip: true
 
         onClicked: {
-            var pos = mapToItem(flickableContents, mouse.x, mouse.y);
-            if (!flickableContents.contains(pos)) {
+            var pos = mapToItem(scrollView.contentItem, mouse.x, mouse.y);
+            if (!scrollView.contentItem.contains(pos)) {
                 root.close();
             }
         }
-- 
cgit v1.1

