diff --git a/src/platforms/xcb/kxmessages.cpp b/src/platforms/xcb/kxmessages.cpp
--- a/src/platforms/xcb/kxmessages.cpp
+++ b/src/platforms/xcb/kxmessages.cpp
@@ -138,7 +138,7 @@
     bool nativeEventFilter(const QByteArray &eventType, void *message, long *result) Q_DECL_OVERRIDE
     {
         Q_UNUSED(result);
-        if (eventType != "xcb_generic_event_t") {
+        if (eventType[0] != 'x') {
             return false;
         }
         xcb_generic_event_t *event = reinterpret_cast<xcb_generic_event_t *>(message);

